### PR TITLE
SNOW-1433003 raise NotImplementedError for Index APIs

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1433003

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Raising NotImplementedError for Index APIs that have not been implemented. I also added documentation for them, and it should be tested.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 1]
```diff
@@ -140,6 +142,40 @@ def values(self) -> ArrayLike:
         """
         return self.to_pandas().values
 
+    @property
+    @index_not_implemented()
```
[the remark this answers is not in the document]
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:147]: would you prefer I do that instead of use the decorator?

[the remark this answers is not in the document]
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:147]: I did it this way because I can then add all APIs to the Index documentation (it will have build errors if specified in the documentation but no method exists in the code).

[the remark this answers is not in the document]
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:147]: We can do both. For those APIs that will be added soon, let's keep what you have done. For those APIs not in our plan yet, we can use the `__getattr__` to raise the error so we can track them in our telemetry too.

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 2]
```diff
@@ -103,6 +136,34 @@ def __init__(
...
+        Returns
+        -------
+        Any
+
+        Notes
+        -----
+        This method also helps raise NotImplementedError for APIs out
+        of current scope that are not implemented.
+        """
+        # TODO: SNOW-1063347: Modin upgrade - modin.pandas.Series functions
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:157]:
```suggestion
```

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 3]
```diff
@@ -103,6 +136,34 @@ def __init__(
...
+        try:
+            return object.__getattribute__(self, key)
+        except AttributeError as err:
+            if key in INDEX_METHODS_NOT_IMPLEMENTED:
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:161]: You don't need this list, you can create a pandas.Index([]) object and check whether the key is an attribute of it, If it is, then raise NotImplementedError. Does that make sense?

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 4]
```diff
@@ -231,43 +329,23 @@ def dtype(self) -> DtypeObj:
         >>> idx.dtype
         dtype('int64')
         """
+        # TODO: SNOW-1458123 implement dtype
         self.to_pandas_warning()
         return self.to_pandas().dtype
 
-    def astype(self, dtype: Any, copy: bool = True) -> Index:
+    @property
+    def inferred_type(self) -> None:
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:337]: add @index_not_implemented()?

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 3 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:159]: Makes sense but would that not cause code coverage issues if I do it that way without getting rid of all the methods below? I can get rid of them if you think that's better

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 4 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:335]: This is being used somewhere in the tests

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 3 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:159]: If the methods is in our plan, we can keep them. Otherwise, we can remove them.

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 4 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:335]: but this method has no implementation here

Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:335]: I don't remember what test was failing but I'll add the decorator

Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:392]: `test_series_setitem_with_empty_key_and_empty_item_negative` fails if it is raising not implemented error

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 5]
```diff
@@ -296,7 +296,7 @@ def dropna():
 
         Keep only the rows with at least 2 non-NA values.
 
-        >>> df.dropna(thresh=2)
+        >>> df.dropna(thresh=2)  # doctest: +SKIP
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:299]: Why is this change necessary?

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 6]
```diff
@@ -425,6 +487,7 @@ def _set_names(self, values: list) -> None:
         WarningMessage.index_to_pandas_warning("_set_names")
         self.to_pandas()._set_names(values)
 
+    # TODO: SNOW-1458122 implement names
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:490]: There's nothing to really do here right once `_set_names` and `_get_names` is implemented?

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 7]
```diff
@@ -466,6 +530,21 @@ def set_names(
             )
         return self.to_pandas().set_names(names, level=level, inplace=inplace)
 
+    @property
+    def ndim(self) -> int:
+        """
+        Number of dimensions of the underlying data, by definition 1.
+        """
+        return 1
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:538]: If only they could all be this easy :)

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 8]
```diff
@@ -474,6 +553,154 @@ def nlevels(self) -> int:
...
+            A single element array-like may be converted to bool.
+
+        See Also
+        --------
+        Index.all : Return whether all elements are True.
+        Series.all : Return whether all elements are True.
+
+        Notes
+        -----
+        Not a Number (NaN), positive infinity and negative infinity
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:637]: TIL NaN stands for Not a Number

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 9]
```diff
@@ -778,6 +1576,7 @@ def tolist(self) -> list:
         >>> idx.to_list()
         [1, 2, 3]
         """
+        # TODO: SNOW-1458117 implement to_list
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:1579]:
```suggestion
        # TODO: SNOW-1458117 implement tolist
```

**Review by sfc-gh-nkrishna — COMMENTED** [review]
Thanks for creating all these docstrings -- left a few comments. Can we also add an index.rst (okay if we do in a follow-up PR)?

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 5 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:299]: ah it was failing previously but no longer does. Good catch!

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 6 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:490]: I think so? I'm not reallu sure what the scope of the task is, I just put JIRA tickets in to make it easier to reference.

Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:490]: this line is not needed.

**Comment by sfc-gh-vbudati** [comment]
@sfc-gh-nkrishna there is a separate PR for documentation here: https://github.com/snowflakedb/snowpark-python/pull/1751
I want to get this one in before since the doc builder is not happy that the functions do not exist lol

**Review by sfc-gh-azhan — APPROVED** [review]

**Review by sfc-gh-nkrishna — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 10]
```diff
@@ -149,6 +153,35 @@ def check_lazy(*args: Any, **kwargs: Any) -> Any:
...
+        -----
+        This method also helps raise NotImplementedError for APIs out
+        of current scope that are not implemented.
+        """
+        try:
+            return object.__getattribute__(self, key)
+        except AttributeError as err:
+            if not key.startswith("_"):
+                native_index = native_pd.Index([])
+                if hasattr(native_index, key):
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:179]: @sfc-gh-vbudati can you add a quick comment here that "any methods that not supported by the current Index.py but exists in native pandas index object should raise a not implemented error for now"

File: src/snowflake/snowpark/modin/plugin/_internal/index.py [hunk 11]
```diff
@@ -312,27 +383,17 @@ def dtype(self) -> DtypeObj:
...
-    def T(self, *args: Any, **kwargs: Any) -> Index:
+    @index_not_implemented()
+    def shape(self) -> None:
         """
-        Return the transpose, which is by definition self.
-
-        Parameters
-        ----------
-        *args : Any
-            Optional positional arguments for compatibility with other T APIs.
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/_internal/index.py:326]: is this not supported any more?

**Review by sfc-gh-yzou — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 11 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:326]: It just got moved further down (see Line 564)

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:326]: I see. Thanks!

File: src/snowflake/snowpark/modin/plugin/_internal/index.py (hunk 10 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:179]: ah I missed this - I'll add it in the documentation PR here: https://github.com/snowflakedb/snowpark-python/pull/1751

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/_internal/index.py:179]: Thanks!
